skytap/client.go

```diff
@@ -188,40 +212,91 @@ func (c *Client) newRequest(ctx context.Context, method, path string, body inter
 	return req, nil
 }
 
-func (c *Client) do(ctx context.Context, req *http.Request, v interface{}) (*http.Response, error) {
-	var resp *http.Response
-	var err error
-	var makeRequest = true
+func (c *Client) do(ctx context.Context, req *http.Request, v interface{}, state *environmentVMRunState, payload responseComparator) (*http.Response, error) {
+	if req.Method == http.MethodPost || req.Method == http.MethodPut || req.Method == http.MethodDelete {
+		for i := 0; i < c.retryCount; i++ {
+			err := c.checkResourceStateUntilSatisfied(ctx, req, state)
+			if err != nil {
+				return nil, err
+			}
+			resp, retry, err := c.requestPutPostDelete(ctx, req, state, payload, v)
+			if !retry || err != nil {
+				return resp, err
+			}
+			waitForAwhile("POST PUT DELETE main loop", "", "", c.retryAfter, i)
+		}
+	}
+	return c.requestGet(ctx, req, v)
+}
 
-	for i := 0; i < c.retryCount+1 && makeRequest; i++ {
-		resp, err = c.hc.Do(req.WithContext(ctx))
+func (c *Client) requestGet(ctx context.Context, req *http.Request, v interface{}) (*http.Response, error) {
+	log.Printf("[DEBUG] SDK GET request (%#v)\n", spew.Sdump(req))
+	resp, err := c.hc.Do(req.WithContext(ctx))
 
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode == http.StatusOK {
+		err := readResponseBody(resp, v)
 		if err != nil {
-			break
+			return nil, err
 		}
+	} else {
+		return resp, c.buildErrorResponse(resp)
+	}
 
-		err = c.checkResponse(resp)
+	return resp, nil
+}
 
-		if err == nil {
-			errBody := readResponseBody(resp, v)
-			if errBody != nil {
-				break
+func (c *Client) requestPutPostDelete(ctx context.Context, req *http.Request, state *environmentVMRunState, payload responseComparator, v interface{}) (*http.Response, bool, error) {
+	var resp *http.Response
+	var err error
+
+	log.Printf("[DEBUG] SDK POST PUT DELETE request (%#v)\n", spew.Sdump(req))
```

Comment:
The request info already provides the verb, probably can be extracted from the request.